apps/src/lib/tools/jsdebugger/redux.js

```diff
@@ -0,0 +1,261 @@
+      jsInterpreter.onPause,
+      () => dispatch(togglePause())
+    );
+    observer.observe(
+      jsInterpreter.onExecutionWarning,
+      (output) => dispatch(appendLog(output))
+    );
+
+    const watchIntervalId = setInterval(
+      () => {
+        const jsInterpreter = getJSInterpreter(getState());
+        getState().watchedExpressions.forEach(we => {
+          const currentValue = jsInterpreter.evaluateWatchExpression(we.get('expression'));
+          dispatch(updateWatchExpressions(we.get('expression'), currentValue));
+        });
+      },
+      WATCH_TIMER_PERIOD
+    );
+    dispatch(clearLog());
+    dispatch({
+      type: ATTACH,
+      observer,
+      watchIntervalId,
+      jsInterpreter,
+    });
+  };
+}
+
+export function detach() {
+  return (dispatch, getState) => {
+    const observer = getObserver(getState());
+    observer && observer.unobserveAll();
+    clearInterval(getWatchIntervalId(getState()));
+    dispatch({type: DETACH});
+  };
+}
+
+export function togglePause() {
+  return (dispatch, getState) => {
+    const jsInterpreter = getJSInterpreter(getState());
+    if (!jsInterpreter) {
+      return;
+    }
+    jsInterpreter.handlePauseContinue();
+    dispatch(mirrorInterpreterState(jsInterpreter));
+  };
+}
+
+export function stepIn() {
+  return (dispatch, getState) => {
+    let jsInterpreter = getJSInterpreter(getState());
+    if (!jsInterpreter) {
+      const runApp = getRunApp(getState());
+      if (runApp) {
+        runApp();
+      } else {
+        throw new Error("jsdebugger has not been initialized yet");
+      }
+      dispatch(togglePause());
+      jsInterpreter = getJSInterpreter(getState());
+      if (!jsInterpreter) {
+        throw new Error("runApp should have attached an interpreter");
+      }
+    }
+    jsInterpreter.handleStepIn();
+    dispatch(mirrorInterpreterState(jsInterpreter));
+  };
+}
+
+export function stepOver() {
+  return (dispatch, getState) => {
+    let jsInterpreter = getJSInterpreter(getState());
+    if (jsInterpreter) {
+      jsInterpreter.handleStepOver();
+      dispatch(mirrorInterpreterState(jsInterpreter));
+    } else {
+      throw new Error("No interpreter has been attached");
+    }
+  };
+}
+
+export function stepOut() {
+  return (dispatch, getState) => {
+    let jsInterpreter = getJSInterpreter(getState());
+    if (jsInterpreter) {
+      jsInterpreter.handleStepOut();
+      dispatch(mirrorInterpreterState(jsInterpreter));
+    } else {
+      throw new Error("No interpreter has been attached");
+    }
+  };
+}
+
+export const actions = {
+  initialize,
+  appendLog,
+  clearLog,
+  attach,
+  detach,
+  togglePause,
+  stepIn,
+  stepOver,
+  stepOut,
+};
+
+// reducer
+
+function appendLogOutput(logOutput, output) {
+  if (logOutput.length > 0) {
+    logOutput += '\n';
+  }
+  if (typeof output !== 'string' && !(output instanceof String)) {
+    output = JSON.stringify(output);
+  }
+
+  logOutput += output;
+  return logOutput;
```

Comment:
rather that storing the log output as a raw string, we should probably store it as an immutable list of objects which get appropriate converted to strings at render time.